core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -648,8 +647,38 @@ private[spark] class BlockManager(
   }
 
   /**
-   * @return true if the block was stored or false if the block was already stored or an
-   *         error occurred.
+   * Retrieve the given block if it exists, otherwise call the provided `makeIterator` method
+   * to compute the block, persist it, and return its values.
+   *
+   * @return either a BlockResult if the block was successfully cached, or an iterator if the block
+   *         could not be cached.
+   */
+  def getOrElseUpdate(
+      blockId: BlockId,
+      level: StorageLevel,
+      makeIterator: () => Iterator[Any]): Either[BlockResult, Iterator[Any]] = {
+    // Initially we hold no locks on this block.
+    doPut(blockId, IteratorValues(makeIterator), level, downgradeToReadLock = true) match {
```

Comment:
I think we can simplify this by moving the acquiring lock logic outside of `doPut`. If we fail to acquire a write lock for the new block we used to return `None` in `doPut` anyway, so it would make sense to instead call `doPut` only if we're able to acquire the write lock. Then you don't need the complicated return type in `doPut`.

(If you don't want to change all the *other* places where `doPut` is called, we can just make an overloaded `doPut` method that takes in a `blockInfo` and assumes the write lock is already held.)